### ISSUE TITLE
page.evaluate() is leaking memory

### ISSUE THREAD

**Issue by jpgklassen** [puppeteer/puppeteer#4059]
### Steps to reproduce

**Tell us about your environment:**

* Puppeteer version: 1.12.2
* Platform / OS version: Ubuntu 18.04.2
* URLs (if applicable): https://www.gofundme.com/cross-country-bike-trip
* Node.js version: 10.15.1

**What steps will reproduce the problem?**

I am using Puppeteer to scrape data from tens of thousands of URLs which are listed in a file. After a few hundred URLs, my program always crashes with this error:

```
Error: Failed to scrape URL https://www.gofundme.com/cross-country-bike-trip: TimeoutError: Navigation Timeout Exceeded: 30000ms exceeded at Promise.then (/var/www/scraper/node_modules/puppeteer/lib/LifecycleWatcher.js:143:21) -- ASYNC -- at Frame. (/var/www/scraper/node_modules/puppeteer/lib/helper.js:108:27) at Page.goto (/var/www/scraper/node_modules/puppeteer/lib/Page.js:662:49) at Page. (/var/www/scraper/node_modules/puppeteer/lib/helper.js:109:23) at scanURLs (/var/www/scraper/middleware/scan-urls.js:30:33) at process._tickCallback (internal/process/next_tick.js:68:7) at scanURLs (/var/www/scraper/middleware/scan-urls.js:205:13)
```

If I restart the program, it always picks up where it left off and scrapes the URL just fine... so I know it is not this particular URL that is the issue.

**What is the expected result?**

Since it scraped hundreds of URLs just fine, and there doesn't appear to be anything special about this URL, and it scrapes this URL just fine when I restart the program, I would expect it to scrape this URL (and all other URLs) without any issues.

**What happens instead?**

The program crashes with the above error.

**My code**

It is very long, but I will include what I believe are the relevant parts here:

index.js:
```
require('dotenv').config()
const fs = require('fs')
const error = require('./middleware/error')
const mongoose = require('mongoose')
const puppeteer = require('puppeteer')
const Marker = require('./models/marker')
const scanSubmaps = require('./middleware/scan-submaps');

(async () => {

  // connect to the database
  await mongoose.connect(
    'mongodb://' + process.env.MONGO_USER_DEV + ':' + process.env.MONGO_PW_DEV + '@' + process.env.MONGO_IP_DEV + ':' + process.env.MONGO_PORT_DEV + '/' + process.env.MONGO_DB_DEV,
    {
      useNewUrlParser: true
    }
  )

  // initiate page
  let page = null
  const browser = await puppeteer.launch({
    headless: true,
    args: ['--full-memory-crash-report']
  })
  page = await browser.newPage()
  await page.setViewport({ width: 1366, height: 768 })
  await page.setRequestInterception(true)

  // don't download unnecessary resources
  page.on('request', (req) => {
    if(req.resourceType() === 'image' || req.resourceType() === 'stylesheet' || req.resourceType() === 'font'){
      req.abort()
    } else {
      req.continue()
    }
  })
  
  // get markers from database
  let markers = await Promise.all([
    Marker.findOne({ name: 'submap' })
      .select('value')
      .exec(),
    Marker.findOne({ name: 'url' })
      .select('value')
      .exec()
  ])
  let submapMarker = markers[0]
  let urlMarker = markers[1]

  // if no submap marker
  if (submapMarker == null) {

    // set submap marker to 0
    submapMarker = 0

  // else, set to its value
  } else {
    submapMarker = submapMarker.value
  }

  // if no URL marker
  if (urlMarker == null) {

    // set URL marker to 0
    urlMarker = 0

  // else, set to its value
  } else {
    urlMarker = urlMarker.value
  }

  // get list of submaps from sitemap
  let sitemap = fs.readFileSync('./sitemaps/sitemap.xml')
  let submapList = String(sitemap).match(/<loc>([^<]+)<\/loc>/g)

  // scan submaps (starting at submap marker)
  await scanSubmaps(page, submapList, submapMarker, urlMarker)

  // delete all markers
  await Marker.deleteMany({})
    .exec()

})().catch(err => {
  await mongoose.disconnect()
  if (typeof page !== 'undefined') {
    await page.close()
  }
  if (typeof browser !== 'undefined') {
    await browser.close()
  }
  error(err)
})
```

scan-submaps.js:
```
const Marker = require('../models/marker')
const fs = require('fs')
const scanURLs = require('./scan-urls')

module.exports = async function scanSubmaps(page, submapList, submapMarker, urlMarker) {

  // slice submapList
  submapList = submapList.slice(submapMarker)

  // for each submap
  for (const item of submapList) {

    // get submap name
    let submapName = item.match(/sitemap[0-9a-z_]+\.xml/)

    // get list of URLs from submap
    let submapContents = fs.readFileSync('./sitemaps/' + submapName)
    let urlList = String(submapContents).match(/<url>(.+?)<\/url>/g)

    // scan URLs (starting at URL marker)
    await scanURLs(page, urlList, urlMarker, submapMarker)

    // increment submap marker
    submapMarker++
    await Marker.findOneAndUpdate({ name: 'submap' }, { value: submapMarker }, { upsert: true })
      .exec()

    // reset URL marker
    urlMarker = 0
    await Marker.findOneAndUpdate({ name: 'url' }, { value: 0 })
      .exec()
  }

  // scan complete
  return 'Submaps scanned successfully.'
}
```

scan-urls.js:
```
const Entry = require('../models/entry')
const Marker = require('../models/marker')

module.exports = async function scanURLs(page, urlList, urlMarker, submapMarker) {

  // slice urlList
  urlList = urlList.slice(urlMarker)
  
  // for each URL
  for (const item of urlList) {

    // get URL
    let url = item.match(/<loc>([^<]+)<\/loc>/)[1]

    try {

      // load URL
      let response = await page.goto(url)

      // if redirect, skip URL
      let redirects = response.request().redirectChain()
      if (redirects.length > 0) {
        // skip URL
      } else {

        // scrape page content
        let scraped = await page.evaluate(() => {

          // SCRAPE A BUNCH OF DATA

          // return scraped data
          return {
            data1: data1,
            data2: data2,
            data3: data3,
            etc: etc
          }
        })

        let entry = {
          url: url,
          data1: scraped.data1,
          data2: scraped.data2,
          data3: scraped.data3,
          etc: scraped.etc
        }

        // add/update database entry
        await Entry.findOneAndUpdate(
          { url: url },
          entry,
          { upsert: true }
        )
          .exec()
      }

      // increment URL marker
      urlMarker++
      await Marker.findOneAndUpdate({ name: 'url' }, { value: urlMarker }, { upsert: true })
        .exec()

    // log URL if error
    } catch (err) {
      throw new Error('Failed to scrape URL ' + url + ': ' + err.stack)
    }
  }

  // scan complete
  return 'URLs scraped successfully.'
}
```

**Comment by jpgklassen** [comment]
**Update:** I edited my code so that it catches the TimeoutError and sends me an email to let me know. But when the program attempts to email me, it fails. This is what I saw in the log file:

```
Mon Feb 25 2019 11:15:10 GMT-0800 (Pacific Standard Time) Error: Failed to deliver email to undefined: Error: spawn ENOMEM
    at transporter.sendMail (/var/www/scraper/middleware/send-email.js:30:18)
    at transporter.send.args (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:226:21)
    at callback (/var/www/scraper/node_modules/nodemailer/lib/sendmail-transport/index.js:88:28)
    at SendmailTransport.send (/var/www/scraper/node_modules/nodemailer/lib/sendmail-transport/index.js:111:20)
    at _processPlugins.err (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:214:34)
    at Mail._processPlugins (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:262:20)
    at _processPlugins.err (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:184:18)
    at processPlugins (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:275:28)
    at err (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:283:17)
    at Mail._convertDataImages (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:387:20)
    at Mail._defaultPlugins.compile.args (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:31:41)
    at processPlugins (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:279:13)
    at Mail._processPlugins (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:287:9)
    at Mail.sendMail (/var/www/scraper/node_modules/nodemailer/lib/mailer/index.js:164:14)
    at /var/www/scraper/middleware/send-email.js:28:19
    at new Promise (<anonymous>)
```

Based on the ENOMEM, it looks like the program has run out of memory and that's why it timed out. I have the server monitoring memory usage every minute, and memory usage is slowly increasing over time as the program runs. This indicates a memory leak, however I have scoured my code for memory leaks and found none (maybe you guys can find one?)

**Comment by jpgklassen** [comment]
**Update:** The program crashed again with the same error. This time, my monitoring caught a spike in memory usage in the same minute as the crash: over 91% of memory was in use. Memory usage in the previous minute was only 73%. Why would it suddenly spike like that?

**Comment by jpgklassen** [comment]
**Update:** After a ton of troubleshooting, I've determined that the memory leak is happening in `page.evaluate()`. When this function returns, every variable declared within the function remains in memory. The callback function that was passed to `page.evaluate()` is also stored in memory, in the form of a string! It makes no difference if I call `page.close()` and `browser.close()` after every single `page.evaluate()`... the variables still continue to accumulate in memory until the program runs out of memory and crashes. I've even tried manually running the garbage collector after every `page.evaluate()`, but it doesn't get rid of the variables.

**This is a big problem for those of us who are using puppeteer for web scraping!!**

**Comment by aslushnikov** [comment]
> I've determined that the memory leak is happening in `page.evaluate()`

@jpgklassen TL;DR: I didn't find a leak here.

I ran DevTools allocation profiler against a simple script:

```js
const puppeteer = require('puppeteer');

(async () => {
    const browser = await puppeteer.launch();
    const page = await browser.newPage();
    for (let i = 0; i < 10000; ++i) {
      await page.evaluate(i => console.log(i), i);
    }
})();
```
The result **does look** like there's a leak:

![image](https://user-images.githubusercontent.com/746130/53665740-22fb6980-3c21-11e9-96eb-332ba4425855.png)

However, if you run this with a new-enough node (I ran node 11), you'll notice that the retainer is a DevTools debugger.

![image](https://user-images.githubusercontent.com/746130/53665830-65bd4180-3c21-11e9-8523-438b0a4bdcca.png)

This is actually a DevTools bug: `page.evaluate` creates a function object to validate input, and DevTools debugger retains the script source to provide debugging capabilities.We should either disable a debugger or filter out devtools-specific GC roots.

Running the script with a quick fix yields the following trace:
![image](https://user-images.githubusercontent.com/746130/53666193-4ffc4c00-3c22-11e9-979a-375a8e870b22.png)

There are some weird memory allocations here and there, but they have nothing to do with Puppeteer's evaluate.

---

To sum up:
- there's no leak in `page.evaluate()`
- there *is* a bug in DevTools allocation profiler. We'll get it fixed soon (cc @a1ph), thanks for bringing this up!
- you're running out of memory for some other reason; it could be your page re-use policy. It is very important to close the page and create a new one every once in a while; otherwise, Chromium will slowly grow render area for the page's renderer over time.
- can you please check if you have the same behavior on older Puppeteer versions?

**Comment by jpgklassen** [comment]
Thanks so much for your reply @aslushnikov. I _was_ using DevTools in development, so that would explain the many copies of the `page.evaluate()` callback function I was seeing in memory. However, I don't think your answer addresses all the variables I was seeing in memory too. It seems every variable declared in the `page.evaluate()` callback is saved in memory, every time `page.evaluate()` is called; and as far as I can tell the retainer for these variables is not the DevTools debugger.

Also, I am not using DevTools in production and there is a memory leak there as well.

Regarding your suggestion to close the page and create a new one every once in a while - I have tried this, and it does slow down the memory leak a bit, but does not fix it unfortunately.

Something else I've noticed is that when the program crashes due to this error, and PM2 restarts it, it often restarts with WAY higher memory usage... even though the Page object, Browser object, and Mongoose connection were closed (in the `catch{}` block) before the program ended! (see below, around 4am)
![memory-usage](https://user-images.githubusercontent.com/10780140/54169835-9ecf9000-4431-11e9-9411-1254458a4cb6.png)

**Comment by aslushnikov** [comment]
> Regarding your suggestion to close the page and create a new one every once in a while - I have tried this, and it does slow down the memory leak a bit, but does not fix it unfortunately.

@jpgklassen IIUC your chart reports total memory allocation for your VM. Can you have a chart with a breakdown of memory allocations between processes? Since there are many parties involved, I wonder who's fault is this: node's, chrome's, or mongo's.

**Comment by jpgklassen** [comment]
@aslushnikov Yes, my chart shows total memory usage for the VM. Unfortunately I don't know of a way to chart memory usage for Chrome or Mongo, but I can use PM2 Plus to track memory usage for the Node program I think. I have just set it up and I'll let it run for a while...

**Comment by jpgklassen** [comment]
**Update:** I turned off PM2 temporarily in order to allow the program to crash, so that I could examine memory usage after the crash. Sure enough, the baseline memory usage after the crash is much higher:
![memory-usage2](https://user-images.githubusercontent.com/10780140/54531328-3c4f2600-4942-11e9-9f78-6ef8c19d7925.png)

After the program crashed, when I examine the process list with the `top` command, there were no processes to account for the additional memory usage:
![top](https://user-images.githubusercontent.com/10780140/54531426-7ae4e080-4942-11e9-8103-e837de70f162.png)

This leads me to believe that the program left behind some stuff in memory. The longer I leave it running with PM2, the closer together the crashes get (because it keeps leaving stuff in memory so it runs out of memory faster and faster). You would think that memory would be cleared when the Node program crashes, but for some reason it isn't... any ideas why?

**Comment by aslushnikov** [comment]
> You would think that memory would be cleared when the Node program crashes, but for some reason it isn't... any ideas why?

@jpgklassen This should never be the case. 

> After the program crashed, when I examine the process list with the top command

This is an in-direct way to inspect system. These observations *might* or *might not* help to determine the root issue. There are [ways](https://unix.stackexchange.com/questions/554/how-to-monitor-cpu-memory-usage-of-a-single-process) to track memory usage per-process in Linux; I'd try using some of these to get a name of a misbehaving service and then file a bug against it to the bugtracker.

I'll close this for now since I didn't find any leak in Puppeteer code per se.

**Comment by jpgklassen** [comment]
@aslushnikov, would you mind commenting on this statement I made earlier:

> However, I don't think your answer addresses all the variables I was seeing in memory too. It seems every variable declared in the page.evaluate() callback is saved in memory, every time page.evaluate() is called; and as far as I can tell the retainer for these variables is not the DevTools debugger.

Do you think this could be the cause of the memory leak?

**Comment by ghost** [comment]
I have the same problem.

So far I’m thinking of restarting the nodejs every hour .... but what to do .... if even browser.close does not solve the leak problem

**Comment by cartene** [comment]
Hi! I have the same problem. I have to launch a different chrome for each webpage I am scraping, because I am using a different proxy for each page. Since I am using docker container I am unable to force the OS (Ubuntu LTS) to clean up the cache or residual memory. Any updates?

**Comment by hatemalimam** [comment]
Same issue here...

**Comment by HoseinGhanbari** [comment]
Same issue here...

**Comment by olegthelilfix** [comment]
Same issue here...

**Comment by rizkinurbarokah** [comment]
I encountered a problem about time scheduling in a repeat for several url pages once or alternately, but he could not run as it should be, various methods have been taken and finally can run only by using an import and export module function and then running with the cron node , but a new problem arises because he cannot use the browser.close () function, but can only use page.close which results in a memory leak because when I use browser.close () then the next process he will not be able to open the browser in next process, is there anyone who can help

**Comment by patryk-zielinski93** [comment]
I have the same problem in Puppeteer 8.0.0. Evaluate function is kept in memory as strings.
![image](https://user-images.githubusercontent.com/19761334/114548328-e869b500-9c5f-11eb-9f27-8e8e6635150f.png)

> We should either disable a debugger or filter out devtools-specific GC roots.

I run page.evaluate a lot of times. How to disable debugger to prevent that strange behavior?

**Comment by Phviana** [comment]
Same issue here...
 Any updates?

**Comment by teo523** [comment]
same issue here...why has this issue been closed?

**Comment by vdkkia** [comment]
+1

**Comment by teo523** [comment]
Hi @aslushnikov, it seems @jpgklassen is right in her comment. Could you please address this?

> @aslushnikov, would you mind commenting on this statement I made earlier:
> 
> > However, I don't think your answer addresses all the variables I was seeing in memory too. It seems every variable declared in the page.evaluate() callback is saved in memory, every time page.evaluate() is called; and as far as I can tell the retainer for these variables is not the DevTools debugger.
> 
> Do you think this could be the cause of the memory leak?

**Comment by garcia-marco** [comment]
+1

**Comment by Nou4r** [comment]
+

**Comment by david-littlefield** [comment]
+1

**Comment by plungarini** [comment]
+1

**Comment by david-littlefield** [comment]
**Update:** 
My issue was caused by my code, and limited experience. Hopefully, this post can help other people just getting started.

**Problem:**
I created a class that stored the browser and page objects as properties of the class object. The problem was that these properties were being shared by visitors instead of created for each visitor. This resulted in these properties being overwritten whenever the class method was executed before the previous class method execution had finished. This caused puppeteer to open a headless chrome browser without being able to close it because the properties were replaced before it could do so. 

**Solution:**
I've written a self-contained javascript file that visibly demonstrates the cause and resolution of this memory leak.
https://gist.github.com/david-littlefield/b9006cc236aaf67f25bfe54b34182805

**Comment by falkenbach** [comment]
> **Solution:** I've written a self-contained javascript file that visibly shows a simplified cause and resolution of this memory leak. https://gist.github.com/david-littlefield/b9006cc236aaf67f25bfe54b34182805

Not relevant, you not doing any page.eval's so yours case is not a solution

**Comment by falkenbach** [comment]
+1

**Comment by david-littlefield** [comment]
Hi @falkenbach, the example I posted was only demonstrating what had actually caused my memory leak, but I had been using page.evaluate in my own project at the time I wrote the initial post. In hindsight, it would have been better to post the example somewhere else, so I've removed the comment and example. Thanks!

**Comment by mikejackowski** [comment]
why was this ticket closed?

**Comment by shirser121** [comment]
Did someone find a solution?

**Comment by musabgultekin** [comment]
I observed a similar leak in playwright. Which makes me wonder that it might not be specific to puppeteer.
